cmd/kube-controller-manager/app/core.go

```diff
@@ -688,30 +691,64 @@ func startGarbageCollectorController(ctx context.Context, controllerContext Cont
 	for _, r := range controllerContext.ComponentConfig.GarbageCollectorController.GCIgnoredResources {
 		ignoredResources[schema.GroupResource{Group: r.Group, Resource: r.Resource}] = struct{}{}
 	}
+
+	// Run and Sync dependencyGraphBuilder if not already
+	if !controllerContext.GraphBuilder.IsRunning() {
+		go controllerContext.GraphBuilder.Run(ctx)
+		go controllerContext.GraphBuilder.Sync(ctx, discoveryClient, garbagecollector.SyncPeriod)
+	}
+
+	attemptToDelete, attemptToOrphan, absentOwnerCache, eventBroadcaster := controllerContext.GraphBuilder.GetGraphResources()
```

Comment:
Why can't the `NewGarbageCollector` get the values it needs?

Reply:
> Why can't the `NewGarbageCollector` get the values it needs?

Still wondering about this. Seems like the `NewGarbageCollector` could call `graphBuilder.GetGraphResources()` and keep the structure contained.  Unit tests would be able to directly the values if it was necessary.

Reply:
This is still outstanding.  Notice how https://github.com/kubernetes/kubernetes/pull/123724 can resolve it.

Reply:
now resolved